Update compatibility with later versions of pyarrow (up to 0.17.0)

fhoering opened the pull request:
## What changes are proposed in this pull request?

- Remove hdfs3 support because current code fails with pyarrow 0.17.0
"TypeError: connect() got an unexpected keyword argument 'driver'"

- Related arrow commit where hdfs3 is deactivated
(see https://github.com/apache/arrow/commit/4e53749097ba687afd5e000067925def2e2802c9)

- Setting ARROW_PRE_0_15_IPC_FORMAT=1 in tests for spark udf:
https://spark.apache.org/docs/latest/sql-pyspark-pandas-with-arrow.html#compatibiliy-setting-for-pyarrow--0150-and-spark-23x-24x

- Even though the tests have been updated it still keeps working with older pyarrow versions, driver will just take 'libhdfs' as default arg

## How is this patch tested?

Tested on our Hadoop cluster

## Release Notes

### Is this a user-facing change?

- [] No. You can skip the rest of this section.
- [x ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
mlflow now supports all versions of arrow until the current latest version 0.17.0.  In order to maintain newer arrow versions we have dropped support for the hdfs3 driver, libhdfs will be chosen as a driver for HDFS access.

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [ ] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Model Registry
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ x] Python

### How should the PR be classified in the release notes? Choose one:

- [ x] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

fhoering committed:
Update compatibility with later versions of pyarrow (up to 0.17.0)

- Remove hdfs3 support because current code fails with pyarrow 0.17.0
"TypeError: connect() got an unexpected keyword argument 'driver'"

- Related arrow commit where hdfs3 is deactivated
(see https://github.com/apache/arrow/commit/4e53749097ba687afd5e000067925def2e2802c9)

- Setting ARROW_PRE_0_15_IPC_FORMAT=1 in tests for spark udf:
https://spark.apache.org/docs/latest/sql-pyspark-pandas-with-arrow.html#compatibiliy-setting-for-pyarrow--0150-and-spark-23x-24x

- Even though the tests have been updated it still keeps working with older pyarrow versions, driver will just take 'libhdfs' as default arg

tomasatdatabricks commented on docs/source/tracking.rst:
```diff
@@ -652,13 +652,7 @@ There are also two ways to authenticate to HDFS:
 Most of the cluster contest settings are read from ``hdfs-site.xml`` accessed by the HDFS native
 driver using the ``CLASSPATH`` environment variable.
 
-Optionally you can select a different version of the HDFS driver library using:
```
Can we still leave this as an option? Just mention that if you use anything else than libhdfs it will not work with newer pyarrow?

fhoering replied:
Thanks for the review. 

If we leave the option for hdfs3 it would introduce code like:
```
if pyarrow > 0.16:
    connected = pa.hdfs.connect(host=host or 'default',
                                port=port or 0,
                                user=kerberos_user,
                                kerb_ticket=kerb_ticket,
                                extra_conf=extra_conf)
else:
    connected = pa.hdfs.connect(host=host or 'default',
                                port=port or 0,
                                user=kerberos_user,
                                driver=driver,
                                kerb_ticket=kerb_ticket,
                                extra_conf=extra_conf)
```

It would make the code less clean.

Is anyone using hdfs3 for real ? What about removing and if somebody complains discuss again to eventually put it back ?

At Criteo we were using hdfs3 with https://github.com/dask/hdfs3 for a long time. It is pretty nice but it is unmaintained and doesn't support viewfs for example. 

I also made some tests and the throughput of libhdfs vs hdfs3 is mostly the same.
The only reason for hdfs is that it doesn't need the JVM mounted via JNI. So 'time to first byte' when starting a new python interpreter is better. But that is really is the only reason.

So we decided to move to libdhfs via arrow's implementation.

In a next PR I would also like to commit https://github.com/apache/arrow/pull/5957 to add viewfs support to mlflow (which only works for libhdfs and not for hdfs3)

Also nobody noticed so far that hdfs3 is broken in later pyarrow versions.

tomasatdatabricks replied:
Those are good points. Can you include a summary of this conversation in the release notes section of the PR description? I.e. say something like - in order to maintain newer pyarrow version we have dropped support for hdfs3 driver.

tomasatdatabricks commented on mlflow/store/artifact/hdfs_artifact_repo.py:
```diff
@@ -164,15 +164,13 @@ def hdfs_system(host, port):
     """
     import pyarrow as pa
 
-    driver = os.getenv('MLFLOW_HDFS_DRIVER') or 'libhdfs'
```
can we still look if the environment is set and if it's set to something other than libhdfs? If that is the case, can we check the pyarrow version and raise an exception if it's incompatible?

tomasatdatabricks commented on tests/pyfunc/test_spark.py:
```diff
@@ -52,14 +49,21 @@ def configure_environment():
-@pytest.fixture
-def spark():
+def get_spark_session(conf):
+    os.environ["ARROW_PRE_0_15_IPC_FORMAT"] = "1"
```
Can you clarify why is this needed in a comment?

tomasatdatabricks replied:
Awesome, thanks!

tomasatdatabricks reviewed (commented):
Hi @fhoering, thanks for the contribution!

Can we still leave an option of setting hdfs driver for people with older pyarrow?

Other than that it looks great!

tomasatdatabricks reviewed (commented):
Looks generally good. 
 - There is still a question I left on the code about why do we need to set os.environ["ARROW_PRE_0_15_IPC_FORMAT"] = "1" in a test. Can you add a comment? 
- Can you add to the release notes a sentence about dropping support for non standard hdfs drivers?

After those 2 points are addressed we can merge. Thanks!

fhoering committed:
Add comment for setting ARROW_PRE_0_15_IPC_FORMAT

fhoering commented:
>     There is still a question I left on the code about why do we need to set os.environ["ARROW_PRE_0_15_IPC_FORMAT"] = "1" in a test. Can you add a comment?

Done

>     Can you add to the release notes a sentence about dropping support for non standard hdfs drivers?
I added in the desription:
`mlflow now supports all versions of arrow until the current latest version 0.17.0. In order to maintain newer arrow versions we have dropped support for the hdfs3 driver, libhdfs will be chosen as a driver for HDFS access.`

And i've put 'breaking-change' but feel free to reclassify.

tomasatdatabricks commented:
Awesome, thank you @fhoering !

tomasatdatabricks commented:
I have restarted the failed Sagemaker test, will merge after it's green.

tomasatdatabricks reviewed (approved):
lgtm!

fhoering commented:
Thanks
